cmd/root/results.go

```diff
@@ -27,28 +28,105 @@ func requiredRunes(s string) []rune {
 	return arr
 }
 
-func wpm(events []event) float64 {
+// Remove backspaces from a list of events.
+//
+// Removing backspace events simplifies some calculations,
+// including wpm,
+func removeBackspaces(events []event) []event {
+	enb := []event{}
+	for _, e := range events {
+		if e.typed != "backspace" {
+			enb = append(enb, e)
+		}
+	}
+	return enb
+}
+
+// Calculates the words per minute (wpm) using the events in the list.
+// Also allows the duration to be overridden, which is useful
+// for calculating wpm per second, which is used in the `wpmGraph` function.
+func wpmBase(e []event, raw bool, d time.Duration) float64 {
+	events := removeBackspaces(e)
+	// cannot calculate wpm with less than 2 events
+	if len(events) < 2 {
+		return 0.0
+	}
 	start := events[0].ts
 	end := events[len(events)-1].ts
-	duration := end.Sub(start)
-	mins := float64(duration) / float64(time.Minute)
+	if d == 0 {
+		d = end.Sub(start)
+	}
+	mins := d.Minutes()
 	wordSize := 5.0
-	chars := events[len(events)-1].i + 1
+	chars := len(events)
 	words := float64(chars) / wordSize
-	incorrect := float64(numIncorrect(events))
-	return (words - incorrect) / mins
+	var result float64
+	if raw {
+		result = (words) / mins
+	} else {
+		incorrect := float64(numUncorrectedErrors(events))
+		if words-incorrect < 0 {
+			return 0.0
+		}
+		result = (words - incorrect) / mins
+	}
+	return result
+}
+
+// Calculates the words per minute (wpm) based on the
+// events that are passed into the
+func wpm(events []event) float64 {
+	return wpmBase(events, false, 0)
+}
+
+// Words per minute per second. Used to calculate the wpm of an
+// array of events that occur within the same second of each other.
+func wpmPs(events []event) float64 {
+	return wpmBase(events, false, time.Second)
 }
 
 // Same as wpm, but doesn't subtract incorrect chars.
 func wpmRaw(events []event) float64 {
-	start := events[0].ts
-	end := events[len(events)-1].ts
-	duration := end.Sub(start)
-	mins := float64(duration) / float64(time.Minute)
-	wordSize := 5.0
-	chars := events[len(events)-1].i + 1
-	words := float64(chars) / wordSize
-	return words / mins
+	return wpmBase(events, true, 0)
+}
+
+func wpmRawPs(events []event) float64 {
+	return wpmBase(events, true, time.Second)
+}
+
+func wpmGraph(events []event) string {
+	d := events[len(events)-1].ts.Sub(events[0].ts)
+	seconds := int(d.Seconds()) + 1
+	wpmData := make([]float64, seconds)
+	wpmRawData := make([]float64, seconds)
+	eventBuckets := make([][]event, seconds)
+
+	for _, event := range events {
+		tsDiff := event.ts.Sub(events[0].ts)
+		bucketId := int(tsDiff.Seconds())
+		eventBuckets[bucketId] = append(eventBuckets[bucketId], event)
+	}
+
+	currSeconds := time.Second
+	var currEvents []event
+	for i, eventBucket := range eventBuckets {
+		// Need to calculate wpm from 0 to i seconds
+		// duration needs to be i + 1 seconds long
+		currEvents = append(currEvents, eventBucket...)
+		wpmData[i] = wpmBase(currEvents, false, currSeconds)
```

Comment:
Don't use wpmBase, but create another function called something like wpmForNSeconds to handle this.